hashing/sha256.cpp

```diff
@@ -0,0 +1,242 @@
+        0x0FC19DC6, 0x240CA1CC, 0x2DE92C6F, 0x4A7484AA, 0x5CB0A9DC, 0x76F988DA,
+        0x983E5152, 0xA831C66D, 0xB00327C8, 0xBF597FC7, 0xC6E00BF3, 0xD5A79147,
+        0x06CA6351, 0x14292967, 0x27B70A85, 0x2E1B2138, 0x4D2C6DFC, 0x53380D13,
+        0x650A7354, 0x766A0ABB, 0x81C2C92E, 0x92722C85, 0xA2BFE8A1, 0xA81A664B,
+        0xC24B8B70, 0xC76C51A3, 0xD192E819, 0xD6990624, 0xF40E3585, 0x106AA070,
+        0x19A4C116, 0x1E376C08, 0x2748774C, 0x34B0BCB5, 0x391C0CB3, 0x4ED8AA4A,
+        0x5B9CCA4F, 0x682E6FF3, 0x748F82EE, 0x78A5636F, 0x84C87814, 0x8CC70208,
+        0x90BEFFFA, 0xA4506CEB, 0xBEF9A3F7, 0xC67178F2};
+
+    const std::string padded_input = prepare_input(input);
+
+    // Process message in successive 512-bit (64-byte) chunks
+    for (size_t i = 0; i < padded_input.length(); i += 64) {
+        std::array<uint32_t, 64> blocks{};
+
+        // Copy chunk into first 16 words of the message schedule array
+        for (size_t j = 0; j < 16; ++j) {
+            blocks[j] =
+                (static_cast<uint8_t>(padded_input[i + j * 4]) << 24) |
+                (static_cast<uint8_t>(padded_input[i + j * 4 + 1]) << 16) |
+                (static_cast<uint8_t>(padded_input[i + j * 4 + 2]) << 8) |
+                static_cast<uint8_t>(padded_input[i + j * 4 + 3]);
+        }
+
+        for (size_t j = 16; j < 64; ++j) {
+            uint32_t s0 = right_rotate(blocks[j - 15], 7) ^
+                          right_rotate(blocks[j - 15], 18) ^
+                          (blocks[j - 15] >> 3);
+            uint32_t s1 = right_rotate(blocks[j - 2], 17) ^
+                          right_rotate(blocks[j - 2], 19) ^
+                          (blocks[j - 2] >> 10);
+            blocks[j] = blocks[j - 16] + s0 + blocks[j - 7] + s1;
+        }
+
+        // Initialize working variables
+        uint32_t a = hash[0];
+        uint32_t b = hash[1];
+        uint32_t c = hash[2];
+        uint32_t d = hash[3];
+        uint32_t e = hash[4];
+        uint32_t f = hash[5];
+        uint32_t g = hash[6];
+        uint32_t h = hash[7];
+
+        // Compression function main loop
+        for (size_t j = 0; j < 64; ++j) {
+            uint32_t s1 =
+                right_rotate(e, 6) ^ right_rotate(e, 11) ^ right_rotate(e, 25);
+            uint32_t ch = (e & f) ^ (~e & g);
+            uint32_t temp1 = h + s1 + ch + k[j] + blocks[j];
+            uint32_t s0 =
+                right_rotate(a, 2) ^ right_rotate(a, 13) ^ right_rotate(a, 22);
+            uint32_t maj = (a & b) ^ (a & c) ^ (b & c);
+            uint32_t temp2 = s0 + maj;
+
+            h = g;
+            g = f;
+            f = e;
+            e = d + temp1;
+            d = c;
+            c = b;
+            b = a;
+            a = temp1 + temp2;
+        }
+
+        // Update hash values
+        hash[0] += a;
+        hash[1] += b;
+        hash[2] += c;
+        hash[3] += d;
+        hash[4] += e;
+        hash[5] += f;
+        hash[6] += g;
+        hash[7] += h;
+    }
```

Comment:
What is your opinion about moving this and the `k` array into a separate function of the type
```cpp
std::array<uint32_t, 8> compute_hash(const std::string& padded_input);
```